packages/http-crawler/src/internals/http-crawler.ts

```diff
@@ -0,0 +1,845 @@
+
+            // Errors are often sent as JSON, so attempt to parse them,
+            // despite Accept header being set to text/html.
+            if (type === APPLICATION_JSON_MIME_TYPE) {
+                const errorResponse = JSON.parse(body);
+                let { message } = errorResponse;
+                if (!message) message = util.inspect(errorResponse, { depth: 1, maxArrayLength: 10 });
+                throw new Error(`${statusCode} - ${message}`);
+            }
+
+            // It's not a JSON, so it's probably some text. Get the first 100 chars of it.
+            throw new Error(`${statusCode} - Internal Server Error: ${body.slice(0, 100)}`);
+        } else if (HTML_AND_XML_MIME_TYPES.includes(type)) {
+            const isXml = type.includes('xml');
+            const parsed = await this._parseHTML(response, isXml, crawlingContext);
+            return { ...parsed, isXml, response, contentType };
+        } else {
+            const body = await concatStreamToBuffer(response);
+            return { body, response, contentType };
+        }
+    }
+
+    protected async _parseHTML(response: IncomingMessage, _isXml: boolean, _crawlingContext: Context): Promise<Partial<Context>> {
+        return {
+            body: await concatStreamToBuffer(response),
+        } as Partial<Context>;
+    }
+
+    /**
+     * Combines the provided `requestOptions` with mandatory (non-overridable) values.
+     */
+    protected _getRequestOptions(request: Request, session?: Session, proxyUrl?: string, gotOptions?: OptionsInit) {
+        const requestOptions: OptionsInit & { isStream: true } = {
+            url: request.url,
+            method: request.method as Method,
+            proxyUrl,
+            timeout: { request: this.navigationTimeoutMillis },
+            sessionToken: session,
+            ...gotOptions,
+            headers: { ...request.headers, ...gotOptions?.headers },
+            https: {
+                ...gotOptions?.https,
+                rejectUnauthorized: !this.ignoreSslErrors,
+            },
+            isStream: true,
+        };
+
+        // Delete any possible lowercased header for cookie as they are merged in _applyCookies under the uppercase Cookie header
+        Reflect.deleteProperty(requestOptions.headers!, 'cookie');
+
+        // TODO this is incorrect, the check for man in the middle needs to be done
+        //   on individual proxy level, not on the `proxyConfiguration` level,
+        //   because users can use normal + MITM proxies in a single configuration.
+        // Disable SSL verification for MITM proxies
+        if (this.proxyConfiguration && this.proxyConfiguration.isManInTheMiddle) {
+            requestOptions.https = {
+                ...requestOptions.https,
+                rejectUnauthorized: false,
+            };
+        }
+
+        if (/PATCH|POST|PUT/.test(request.method)) requestOptions.body = request.payload;
```

Comment:
are we sure the `request.method` is always uppercase?

Reply:
The spec says that the method is case-sensitive. I haven't changed this piece of code, so it was like that before.